sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,450 @@
+  //
+  // The only access pattern intended for Beam, because of its superior
+  // performance possibilities, is "urn:beam:sideinput:multimap" (or some such
+  // URN)
+  string access_pattern = 1;
+
+  // (Required) A UDF that adapts a particular access_pattern to a user-facing
+  // view type.
+  //
+  // For example, View.asSingleton() may include a `view_fn` that adapts a
+  // specially-designed multimap to a single value per window.
+  //
+  // TODO: this will actually be reused a lot; consider making it by reference
+  FunctionSpec view_fn = 2;
+
+  // (Required) A UDF that maps a main input window to a side input window.
+  //
+  // For example, when the main input is in fixed windows of one hour, this
+  // can specify that the side input should be accessed according to the day
+  // in which that hour falls.
+  FunctionSpec window_mapping_fn = 3;
+}
+
+// An environment for executing UDFs. Generally an SDK container URL, but
+// there can be many for a single SDK, for example to provide dependency
+// isolation.
+message Environment {
+
+  // (Required) A pipeline-scoped unique id for this node
+  int64 id = 1;
+
+  // (Required) The URL of a container
+  //
+  // TODO: reconcile with Fn API's DockerContainer structure by
+  // adding adequate metadata to know how to interpret the container
+  string url = 2;
+}
+
+// Description of a user-defined function, including the environment/SDK in
+// which to execute it
+message FunctionSpec {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A cross-language, stable, unique identifier for the function.
+  //
+  // The `payload` field should be understandable by any entity familiar with
+  // the URN, with no insight into the completely opaque `data` blob. Together,
+  // the URN and the payload should suffice for optimization or language-to-
+  // language replacement.
+  //
+  // For example the distinguished urn "urn:beam:windowfn:FixedWindows" with
+  // payload `{ duration: n }` fully specifies a windowing function which can
+  // be implemented by the SDK constructing the pipeline, by another SDK (for
+  // language-to-language fusion compatibility) or by the runner directly.
+  string urn = 2;
+
+  // (Required) Reference to the specification of the execution environment
+  // required to invoke this function.
+  int64 environment_reference = 3;
+
+  // (Optional) If present, a payload whose schema is determined by the URN of
+  // this function.
+  google.protobuf.Any payload = 4;
+
+  // (Required) The raw data of the function that the SDK knows how to
+  // deserialize, but need not be comprehensible to any other runner, SDK, or
+  // other entity.
+  google.protobuf.Any data = 5;
+}
+
+// TODO: transfer javadoc here
+message DisplayData {
+
+  // Map keys cannot be messages, so the Identifier needs to be flattened to a string
+  map<string, Item> items = 1;
+
+  message Identifier {
+    repeated string path = 1;
```

Comment:
The `path` identifies where the display data originated from within the graph, i.e. a `GraphNode` here. If there's a better way of identifying `GraphNode`s. If we're now keeping the full hierarchy of composites (yay!), this could simply be a `GraphNode` `id`.